p2p/proto/snapshot.proto

```diff
@@ -0,0 +1,109 @@
+syntax = "proto3";
+
+import "p2p/proto/common.proto";
+import "p2p/proto/state.proto";
+
+message PatriciaNode {
+    message Edge {
+        uint32  length = 1;
+        Felt252 path   = 2;  // as bits of left/right
+        Felt252 value  = 3;
+    }
+    message Binary {
+        Felt252 left  = 1;
+        Felt252 right = 2;
+    }
+
+    oneof node {
+        Edge   edge   = 1;
+        Binary binary = 2;
+    }
+}
+
+// non leaf nodes required to build the trie given the range (leaves)
+message PatriciaRangeProof {
+    repeated PatriciaNode nodes = 1;
+}
+
+// leafs of the contract state tree
+message ContractState {
+    Hash   address = 1;  // the key
+    Hash   class   = 2;
+    Hash   storage = 3;  // patricia
+    uint64 nonce   = 4;
+}
+
+// request a range from the contract state tree that matches the given root (block)
+// starts at 'start' and ends no more than 'end'.
+// the result is  (ContractRange+, PatriciaRangeProof)*
+message ContractRangeRequest {
+    uint32 domain           = 1;  // volition
+    Hash   state_root       = 2;
+    Hash   start            = 3;
+    Hash   end              = 4;
+    uint32 chunks_per_proof = 5;  // how many ContractRange items to send before sending a proof
+}
+
+// stream of leaves in the contracts tree
+message ContractRange {
+    repeated ContractState state = 1;
+}
+
+message ContractRangeResponse {
+    Hash state_root     = 1;
+    Hash contracts_root = 2;
+    Hash classes_root   = 3;
+    oneof responses {
+        ContractRange range = 4;
+        Fin           fin   = 5;
+    }
+}
+
+// duplicate of GetContractRange. Can introduce a 'type' instead.
+// result is (Classes+, PatriciaRangeProof)*
+message ClassRangeRequest {
+    Hash   root             = 1;
+    Hash   start            = 2;
+    Hash   end              = 3;
+    uint32 chunks_per_proof = 4;
+}
+
+message ClassRangeResponse {
+    Hash root           = 1;
+    Hash contracts_root = 2;
+    Hash classes_root   = 3;
+    oneof responses {
+        Classes classes = 4;
+        Fin     fin     = 5;
+    }
+}
+
+// A position in some contract's state tree is identified by the state tree's root and the key in it
+message StorageLeafQuery {
+    Hash    contract_storage_root = 1;
```

Comment:
What happen when the [state_root](https://github.com/starknet-io/starknet-p2p-specs/pull/8/files#diff-1895f60c4bbe4eec02ba055cd529fd9ad7cb4b76bad983cbc68a2d1dc0ce0190R95) becomes too old and no longer available from the network? The client will just simply pick a newer state root, but the `contract_storage_root` for this contract may have changed by then, so you'll need to have a way to 'refresh' this contract's root. See [this](https://github.com/NethermindEth/juno/pull/1030/files#diff-a04fd2fd0785454e284c95dea53642c1d266c8cbbf66a8862b330362cf710e87R156)

Reply:
If it is no longer available, how can the server know what `contract_storage_root` refers to?

Reply:
The server can know the correct `contract_storage_root` from `state_root` and `key`. If it does not match, then a refresh mechanism is needed. On ethereum, the server just send nothing and its up to the client to re-fetch the correct `contract_storage_root` from another endpoint.

Reply:
You said the contract_storage_root changed (i'm not sure we're on the same page as to what exactly it is) and in particular the old data no longer exists.

Reply:
Concretely, lets say the block you need to sync against is no longer in the network and you want to sync with a new block. Take the slices you have and run a queries with their most extreme keys against the new block. You will get patricia proofs that allow you to verify if one of the keys within the slices changed. It will also include a small set of values in the response (one per key), which may be a bit redundant, but I don't think it warrants a whole new query type. If we see there's a need to optimize we can introduce a flag into the query so as to only return the patricia proofs.

Reply:
Yea, so the problem is, the proof may produce a new `contract_storage_root` if the contract was changed between the new block and the new block. At this point, we don't know what is the new `contract_storage_root`, because the `contract_storage_root`, is the one we got from contract request before when the block was still available, so we can't verify the storage range yet, so we need to get the new `contract_storage_root` and verify that it belongs to the new block's state root.

Reply:
Right, issue a new ContractRangeRequest to get the slice again and now you have all roots in that slice as they are for that block. In order to know which slices to ask for you can do the procedure I mentioned above.

Reply:
I see, ok that'll work too.